### PR TITLE
08588 Grant Approval system contract

### LINKED ISSUE #8588: GRANT_ALLOWANCE (ApprovePrecompile)

### PR DESCRIPTION
**Pull request by stoyanov-st**
**Description**:
Implementation for the HTS approve and approveNFT functions.

**Related issue(s)**:

Fixes #8588 #8589 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by MrValioBg — APPROVED** [review]
LGTM!

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalTranslator.java
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalTranslator.java:36]: Thanks! ERC approve added.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalTranslator.java
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalTranslator.java:71]: nit: what is the advantage in having this as a separate method?  I guess clarity?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalDecoder.java
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalDecoder.java:88]: Do you know if the amount is ever validated to being not negative for fungible tokens?  I presume it may be done in the token service but it might help to check.

Comment (lukelee-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/grantapproval/GrantApprovalDecoder.java:88]: thanks for checking.

**Review by lukelee-sl — DISMISSED** [review]
LGTM

**Comment by stoyanov-st** [comment]
> We are missing bool output for ERC approve
> `approve(address spender, uint256 amount) → bool`

The ERC approve functions are added

**Comment by tinker-michaelj** [comment]
> > Nice `XTest` 🙌 .
> > Same comment as [here](https://github.com/hashgraph/hedera-services/pull/8856#pullrequestreview-1653007156); I don't think we need a separate `HtsCall` implementation here.
> 
> We have different responses for the ERC functions that do not have `status` value thus the `HtsCall` implementation. Should we try to implement a custom `SingleTransactionRecordBuilder` to match this case and use it with `DispatchForResponseCodeHtsCall`?

Ah that's a very good point. Let's merge this and revisit once we've characterized mono-service behavior. 💯

**Review by tinker-michaelj — APPROVED** [review]
LGTM! 🥂

**Comment by rbair23** [comment]
There are two failing HAPI tests in the ApproveAllowanceSuite. Can you review these before merging? Thanks!

image

**Pull request merged**
